655 Add User Roles and Basic User Data to Db

Joshua-Douglas opened the pull request:
Closes #655.

### What changes did you make?

Added user roles (`Guest`, `Host`, `Coordinator`, `Admin`) and updated the user model to include first, middle, and last name. PR #661 made an update that passes the first and last name to the backend, so I was also able to update the frontend app to begin receiving and storing the user names as well (video below).

The user name initials are now displayed on the user avatar menu after logging in, and the user's name is accessible from all authenticated components.

Making this change required refactoring our alembic migration scripts, the API test project, and the API itself. The following detailed changes were required:
- Add a database version check during API startup. If the database is not at the head revision then an error will be thrown to notify the developer that an alembic migration is required. The logic relies on a static version number that should be incremented.
- Integrate the alembic database migration process into our test client setup fixture. Our backend tests will run against a properly versioned database, to ensure the test database has all pre-populated values (such as the `roles` table). 
- Upgrade `marshmallow-sqlalchemy` to the latest release. I needed a bug fix from the latest release.
- Update `signUpHost` and `signUpCoordinator` endpoints to 1) assign a role to their created users and 2) require the user first and last name.
- Update the `user` and `signIn` endpoints to return the user first and last name as part of the `user` data.
- Update data model diagram within `api/openapi_server/models/README.md`, and include instructions on how to update the database model. Turns out there are a lot of steps!
- Add `User` model (de)serialization json schema, complete with a test suite. We could add some name validation here (e.g. don't allow spaces in names) and filter the json requests through the schema, but it didn't seem worth while.
- Remove the `Host` model object. This is now replaced by the `User` model with a `Host` role.
- Remove the `create_host` endpoint. This is now completely replaced by `signUpHost`
- Modify the base alembic migration script, `3ceec084158f_.py`, to work with unversioned local databases. This will allow other devs to run `alembic upgrade head` even if they've been working off of an API-generated database.
- Add migration script to introduce user roles and user names. All existing users will be assigned the `guest` role and will be given the names "unknown".
- Update `run-tests.yml` to include a `workflow_dispatch` trigger to allow manual triggering.
- Add `cypress:run:mock` and `cypress:run:nomock` configurations to allow more easily running the full e2e test suites.
### Rationale behind the changes?

Before this PR there was no mechanism for determining user type and there was no mechanism for storing basic user information. We can now differentiate between guests, hosts, and coordinators. This will allow us to begin creating different views for each user type, and to begin restricting portions of the API that require elevated access.

This PR does not implement any data model relationships between user types, and it does not implement role-specific user data.

### Testing done for these changes
  - Added a suite of backend alembic migration tests that will test all current and future migration scripts. These tests cycle through the upgrades and downgrade scripts and check for common database migration errors. These tests are assisted by the new dev dependency, `pytest-alembic`.
  - Modified the e2e sign up tests to include a 'no mocking' mode. I ran the signup tests against the real API to confirm our signup still works.
  - Tested each of the modified endpoints
  - Tested the new schema objects, and their validation

### What did you learn or can share that is new?(optional)
 
I learned how to write and test `alembic` database migrations. The steps are documented in `api/openapi_server/models/README.md`. 

### New Data Model
```mermaid
classDiagram
class alembic_version{
   *VARCHAR<32> version_num NOT NULL
}
class housing_program{
   *INTEGER id NOT NULL
   VARCHAR program_name NOT NULL
   INTEGER service_provider NOT NULL
}
class housing_program_service_provider{
   *INTEGER id NOT NULL
   VARCHAR provider_name NOT NULL
}
class role{
   *INTEGER id NOT NULL
   VARCHAR name NOT NULL
}
class user{
   *INTEGER id NOT NULL
   VARCHAR email NOT NULL
   VARCHAR<255> firstName NOT NULL
   VARCHAR<255> lastName NOT NULL
   VARCHAR<255> middleName
   INTEGER role_id NOT NULL
}
housing_program_service_provider "1" -- "0..n" housing_program
role "1" -- "0..n" user
```

### Video of new Sign In User Initials
<details>
<summary>Visuals after changes are applied</summary>
 
https://github.com/hackforla/HomeUniteUs/assets/22138019/9bdf0326-269e-432b-be1f-335f64605367

![image](Paste_Your_Image_Link_Here_After_Attaching_Files)

</details>

erikguntner commented on .github/workflows/run-tests.yml:
```diff
@@ -4,6 +4,7 @@ on:
     branches: [main]
   # This is also a reusable workflow that can be called from other workflows
   workflow_call:
+  workflow_dispatch:
```
What is this doing?

Joshua-Douglas replied:
The `workflow_dispatch` event lets you manually trigger the workflow. Sometimes it is useful to run the remote test cases before submitting a PR, so I usually like to include this.

erikguntner reviewed (approved):
Everything looks great! Thanks for the detailed steps for updating the data model. Approving. Only question is regarding the `workflow_dispatch` property in the test workflow.

erikguntner commented on api/openapi_server/models/database.py:
```diff
@@ -1,14 +1,26 @@
+    middleName = Column(String(255), nullable=True)
+    lastName = Column(String(255), nullable=False)
+    role_id = Column(Integer, ForeignKey('role.id'), nullable=False)
+    role = relationship("Role", back_populates="users")
```
Does this return the role as a string on the user object?

Joshua-Douglas replied:
It returns a `models.database.Role` object. The `relationship` function uses the `role_id` to perform an automatic lookup of the `Role` object whenever the `user` object is accessed. It'll also perform some validation when updating the `role_id` to make sure it is a valid `Role`. 

The `User` table doesn't actually store a `role` - just the `role_id`: 
![image](https://github.com/hackforla/HomeUniteUs/assets/22138019/7db6d020-bfa9-4b36-9e8a-7526f15e394f)

But if you dump the `user` model to json you see the serialized `role` object:
`{'role': {'name': 'Admin'}, 'email': 'testcoordinator@test.com', 'firstName': 'testuser_firstname', 'middleName': '', 'lastName': 'testuser_lastname'}`

erikguntner replied:
Okay cool. If that's the object being returned when the user logs in then that's perfect. Just wanna make sure we have the role on the front end so we can do things like render routes based on role, show different components, etc.

paulespinosa commented on api/alembic/versions/e4c8bb426528_add_user_types.py:
```diff
@@ -0,0 +1,70 @@
+      * There is no way to map host users back to the user table. We would need a user id foreign 
+        key, or at least an email address.
+    '''
+    role_table = op.create_table('role',
```
This means there can be duplicate role names. The `id` column isn't strictly necessary. The `name` column can be the primary key.

Joshua-Douglas replied:
Correct, the `id` isn't strictly necessary, but it does provide us the flexibility to rename the roles at any time. I think being able to easily rename our roles is worth storing 4-10 extra integers in the db.

To your point about duplicate names, I added a safeguard against this in [6ad02dd](https://github.com/hackforla/HomeUniteUs/pull/668/commits/6ad02dd8234cceeb29d77c55fff3045dfaf495c8) by adding the `unique=True` constraint to the `role.name` field.

paulespinosa replied:
Using `on update cascade` can also be considered easy when defining a table. What other reasons are there to use a surrogate key?

Joshua-Douglas replied:
Using surrogate keys like an id field offers several advantages beyond the ease of updates provided by `ON UPDATE CASCADE`. Surrogate keys are unique and stable, providing an abstraction layer that decouples database structure from business logic, thus reducing the need for changes due to business modifications. They enhance performance with quicker indexing and simplify relational management by avoiding the complexities and potential performance issues of cascading updates. Additionally, surrogate keys maintain data integrity by minimizing risks associated with data duplication and erroneous updates, making them ideal for robust, scalable database architectures.

paulespinosa replied:
Is that a chatgpt response? :smile: Do these reasons apply to this data model?

paulespinosa commented:
> * (e.g. don't allow spaces in names)

Some people have spaces in their names. Here's some resources to understand more about names:

* https://designsystem.digital.gov/patterns/create-a-user-profile/name/
* https://shinesolutions.com/2018/01/08/falsehoods-programmers-believe-about-names-with-examples/

https://github.com/kdeldycke/awesome-falsehood

paulespinosa commented:
Looking good. Is there a PM created issue this can relate back to?

Joshua-Douglas commented:
> > * (e.g. don't allow spaces in names)
> 
> Some people have spaces in their names. Here's some resources to understand more about names:
> 
> * https://designsystem.digital.gov/patterns/create-a-user-profile/name/
> * https://shinesolutions.com/2018/01/08/falsehoods-programmers-believe-about-names-with-examples/
> 
> https://github.com/kdeldycke/awesome-falsehood

That's a great illustration of why I did not validate the names. Great sources, thanks!

After reading through them I realized two things and corrected both in [6ad02dd](https://github.com/hackforla/HomeUniteUs/pull/668/commits/6ad02dd8234cceeb29d77c55fff3045dfaf495c8):
1. Last names should not be required. I updated the database model to only require first names.
2. First names should be at least one character in length. The data model now validates that the name is at least one non-space character.

Joshua-Douglas commented:
Hey @erikguntner and @paulespinosa, 

Thank you so much for taking the time to do a detailed review. I've updated the PR in response to your comments. 

Since I had to modify the migration script, it is probably best for you to downgrade and re-upgrade if your local db is on the current head. 

Here are all the changes from the review:
- Make the user last name optional, since some cultures only use one name. I modified both the frontend and backend to accommodate this change.
- Add validation to the user first name, to require at least 1 non-space character in length.
- Auto-increment the database head version by reading from the alembic configuration file. Before this change the version had to be manually incremented, but this change removes a manual step from the db upgrade process.

I also noticed that #602 broke our e2e tests, so I submitted a fix for this in [74f861e](https://github.com/hackforla/HomeUniteUs/pull/668/commits/74f861e4e8bf16b52a9cc5cffa2e0607e23cba9c).

Joshua-Douglas commented:
@paulespinosa, I saw your question but I don't see any requested changes. Does this PR require more work?

paulespinosa commented:
> @paulespinosa, I saw your question but I don't see any requested changes. Does this PR require more work?

Code seems fine. @sanya301 this feature adds fields to the sign up page as shown in the video above. Look good to you?
